zaza/openstack/charm_tests/nova/tests.py

```diff
@@ -68,6 +78,139 @@ def test_launch_small_instance(self):
+
+        # Check that nova-compute services are enabled before testing
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'enabled')
+
+        # Run 'disable' action on units
+        zaza.model.run_action_on_units([unit.name for unit in nova_units],
+                                       'disable')
+
+        # Check action results via nova API
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'disabled')
+
+        # Run 'enable' action on units
+        zaza.model.run_action_on_units([unit.name for unit in nova_units],
+                                       'enable')
+
+        # Check action results via nova API
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'enabled')
+
+    def test_950_remove_from_cloud_actions(self):
+        """Test actions remove-from-cloud and register-to-cloud.
+
+        Note (martin-kalcok): This test requires that nova-compute unit is not
+        running any VMs. If there are any leftover VMs from previous tests,
+        action `remove-from-cloud` will fail.
+        """
+        all_units = zaza.model.get_units('nova-compute',
+                                         model_name=self.model_name)
+
+        unit_to_remove = all_units[0]
+
+        service_name = self.fetch_nova_service_hostname(unit_to_remove.name)
+
+        registered_nova_services = self.nova_client.services.list(
+            host=service_name, binary='nova-compute')
+
+        service_count = len(registered_nova_services)
+        if service_count < 1:
+            self.fail("Unit '{}' has no nova-compute services registered in"
+                      " nova-cloud-controller".format(unit_to_remove.name))
+        elif service_count > 1:
+            self.fail("Unexpected number of nova-compute services registered"
+                      " in nova-cloud controller. Expecting: 1, found: "
+                      "{}".format(service_count))
+
+        # run action remove-from-cloud and wait for the results in
+        # nova-cloud-controller
+        zaza.model.run_action_on_units([unit_to_remove.name],
+                                       'remove-from-cloud',
+                                       raise_on_failure=True)
+
+        # Wait for nova-compute service to be removed from the
+        # nova-cloud-controller
+        sleep_timeout = 1  # don't waste 10 seconds on the first run
+
+        for _ in range(31):
+            sleep(sleep_timeout)
+            service_list = self.nova_client.services.list(
+                host=service_name, binary='nova-compute')
+            if len(service_list) == 0:
+                break
+            sleep_timeout = 10
+        else:
+            self.fail("nova-compute service was not unregistered from the "
+                      "nova-cloud-controller as expected.")
```

Comment:
This code is sadly repeated below, which violates DRY.  There are several ways to make it better; I suspect tenacity would be a good approach.  However, sadly the actual review in the charm has been landed so this PR needs to be landed.  We'll have to fix it later.